plutus-core/docs/equalsCanonical.md

```diff
@@ -0,0 +1,203 @@
+# Equality and serialization built-in functions for SOPs
```

Comment:
> We'd need to have an actual workable size measure on terms

So 2022... :)

All you need to know is what it costs to check equality of two `VConstr` AST nodes and what it costs to check equality of two constants of the given type (all of which can be calibrated separately), then you just stream those individual costs to the outside as you process the arguments to `equalsCanonical`. The only non-obvious part are how we scale the individual costs (do we have a ton of parameters to scale costs for checking equality of integers, of bytestrings etc or do we hardcode a single parameter and pray that it scales all kinds of costs correctly?) and whether we should just reuse the already existing costing functions for `equalsInteger`, `equalsByteString` etc. None of which is new, because we have these same questions for `equalsData` too.

Really, tell me how `equalsCanonical` is that different from `equalsData` that it becomes an absolute nightmare (assuming you don't believe costing of `equalsData` is an absolutely nightmare too, in which case we might have made a bit of a mistake by introducing it).

I really feel like the majority of "costing is complicated" arguments is caused merely by not having the right view at costing. I need to return to [Operational semantics for builtins](https://input-output.atlassian.net/browse/PLT-1007) some day... Particularly given that lazy costing was a huge step in the right direction and we already seem to have all the necessary infrastructure, we just need to shift the view towards costing a bit.

Reply:
> The only non-obvious part are how we scale the individual costs (do we have a ton of parameters to scale costs for checking equality of integers, of bytestrings etc or do we hardcode a single parameter and pray that it scales all kinds of costs correctly?)

Right, and also "how do we describe the costing of this function?". At the moment, all of our costing functions can be described statically as one of a family of simple functions based on the sizes of the inputs. I think that's a nice property!

Reply:
> > The only non-obvious part are how we scale the individual costs (do we have a ton of parameters to scale costs for checking equality of integers, of bytestrings etc or do we hardcode a single parameter and pray that it scales all kinds of costs correctly?)
>
> Right, and also "how do we describe the costing of this function?". At the moment, all of our costing functions can be described statically as one of a family of simple functions based on the sizes of the inputs. I think that's a nice property!

I seriously don't understand how costing of `equalsCanonical` is any different to costing of `equalsData`. It may be in some way, but costing-wise they look identical to me (particularly given that we literally presuppose that any canonical value can be converted to `data` which can subsequently be fed to `equalsData`), so if you can enlighten me on that, I'd appreciate it a lot. 

The part of my comment that you quote ends with

> None of which is new, because we have these same questions for `equalsData` too.

btw.

Reply:
Ah right, so in fact the problematic part of `equalsData` is not the costing function itself, but the _size measure_. The costing function for any such equality function is probably always going to be "minimum of the two sizes".  I do agree that our size measures are much more of a mess than our costing functions. The size measure for `Data` includes the sizes of the internal integers and bytestrings, which is a bit awkward but I think makes some sense for functions that uniformly operate over its structure like equality and serialization (it would be pretty bad if for some reason, say, serializing integers was non-linear!).

Now, coming up with a size measure for an arbitrary value is awkward (functions etc.), but at least somewhat makes sense for a canonical value. That does mean that we would need to rely _semantically_ on lazy costing, since we would need to fail the evaluation if the term is not canonical, so we can't even in principle compute the size and cost up front (which we can still in principle do for `equalsData`, even if the implementation chooses to be more efficient).

The other way in which things are more complicated for `equalsCanonical` is that `Data` contains _known_ other types at leaf nodes, whereas canonical values can contain leaf nodes of arbitrary builtin types. So while we might reason that the costing behaviour of `equalsData` is reasonable, based on the two other types it contains, it's much harder to reason about `equalsCanonical`.